### DISCUSSION TITLE
Compression support in Faraday 2

### DISCUSSION THREAD

**Discussion by bodrovis** [lostisland/faraday#1390] [Help] [answered]
First of all, thank you for supporting this gem, it's so useful! I'm a bit lost and perhaps someone could clarify this for me. I've been using Faraday 1 with faraday_middleware, specifically I've added a plugin that was enabling GZip support. Here's the relevant code: https://github.com/lokalise/ruby-lokalise-api/blob/master/lib/ruby-lokalise-api/connection.rb#L9 It worked fine but now this middleware is deprecated. The documentation states that the GZip plugin is not needed when using NetHttp adapter as it will add the "accept-encoding" header and will automatically decompress the response.

However, I don't see this happening: after I've removed the GZip plugin and updated to Faraday 2 my tests started to fail because the response is not decompressed: `response.body` returns a compressed string. Also, I don't see the "accept-encoding" header being added automatically judging by the recorded VCR cassettes - this header is just not there, and the decompression does not happen for me if I try to add it manually.

Here's a simple example:

```
irb(main):001:0> require 'faraday'
=> true
irb(main):002:0> conn = Faraday.new
=>
#<Faraday::Connection:0x000002bce01f4088
...
irb(main):003:0> conn.adapter
=> Faraday::Adapter::NetHttp
irb(main):004:0> conn.headers
=> {"User-Agent"=>"Faraday v2.1.0"}
irb(main):005:0>
```

Then I'm trying to:

```
resp = conn.get 'http://example.com'
```

When inspecting `resp` I can see the following:

```
request_headers={"User-Agent"=>"Faraday v2.1.0"}
```

So, the only header I see is "user-agent" though I expected to find "accept-encoding" here as well. I'm pretty sure I'm missing a point here. How should it work?

P.S. I've just tried to use HTTPx adapter and it worked fine

**Comment by iMacTia** [comment, marked as answer]
Hi @bodrovis and thanks for reaching out!
This is indeed interesting because `Net::HTTP` [officially supports compression out of the box](https://ruby-doc.org/stdlib-2.1.1/libdoc/net/http/rdoc/Net/HTTP.html#class-Net::HTTP-label-Compression).
I'm honestly unsure as to why this isn't the case for you, but I can only imagine this is due to the lack of some necessary header (that the server should return?) 🤔.

If you'd like to keep supporting all adapters, then the `gzip` could be gemified into its own gem. If you're interested, the process is actually quite easy thanks to the [faraday-middleware-template](https://github.com/lostisland/faraday-middleware-template) repository. The main reason this hasn't been done it is because we didn't consider the middleware much useful anymore, but if someone would like to take the lead in making a gem out of it compatible with Faraday 2.0, then I'd be more than happy to help 🙌

**Reply by bodrovis** [reply]
Yeah, I'm a bit puzzled myself as well because now when I switched to HTTPx everything is working correctly, as it used to be. I tried various approaches with net_http to no avail, not sure, what's going on. I thought about porting gzip for Faraday 2 and I'll research it a bit more. In any case, thank you for a response!
